### PR TITLE
fix(query) SetOperator LOR are not compliant with prometheus

### PR DESCRIPTION
**Pull request by amolnayak311**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

**Current behavior :** 

The prometheus ``OR`` operation for a range query currently looks for existence of join keys in the range and then does not give results for RHS even if the LHS is not available for entire time range in the query range. This behavior is not same as prometheus

Consider ``sum(foo{}) OR vector(1)`` for range ``now-7d`` to ``now``. Suppose ``sum(foo{})`` is only available from  ``now-7d`` to ``now - 1d``, we must see ``vector(1)`` from  ``now-1d`` to ``now``.  Currently. because we check the existence of a join key in the entire range, we don't see ``vector(1)`` in the results for the missing time range. 

Additionally, all binary joins considered absence of ``on`` and assumed that to be empty list. This isn't the right assumption as not providing``on`` and ``on()`` both would give ``on`` as empty list. This PR fixes this by making ``on`` an option where its ``None`` if on isn't specified and ``Some`` if its specified. An example is ``foo{} OR vector(1)`` was treated same as ``foo{} OR on() vector(1)`` before the fix. 

**New behavior :**

Above incompatibility is fixed. 

**Additional notes**

Similar regression in behavior for ``AND`` and ``UNLESS`` too might be present and would need more investigation and a fix on similar lines

### CODE REVIEW

File: query/src/main/scala/filodb/query/exec/SetOperatorExec.scala
Comment (yu-shipit) [inline query/src/main/scala/filodb/query/exec/SetOperatorExec.scala:292]: This part of code is same as the logic for the left. can we have a common function for it?

File: query/src/main/scala/filodb/query/exec/BinaryJoinExec.scala
Comment (amolnayak311) [reply query/src/main/scala/filodb/query/exec/BinaryJoinExec.scala:33]: I'm not sure I follow the question, is the onLabel ``None``? In that case the join key will contain **all** the labels except the ``ignoringLabels``.

File: query/src/main/scala/filodb/query/exec/SetOperatorExec.scala
Comment (amolnayak311) [reply query/src/main/scala/filodb/query/exec/SetOperatorExec.scala:292]: Its not the same, the lhs is **setting** a bit and rhs is **checking** for the set bit and returning a NaN or a value. Am I missing something?

File: query/src/main/scala/filodb/query/exec/SetOperatorExec.scala
Comment (alextheimer) [inline query/src/main/scala/filodb/query/exec/SetOperatorExec.scala:135]: Looks like we shouldn't care whether-or-not the RHS value is `NaN`: http://demo.robustperception.io:9090/graph?g0.expr=vector(0)%20and%20vector(Nan)&g0.tab=1&g0.stacked=0&g0.range_input=1h

File: query/src/main/scala/filodb/query/exec/SetOperatorExec.scala
Comment (alextheimer) [inline query/src/main/scala/filodb/query/exec/SetOperatorExec.scala:275]: It looks like the previous implementation was agnostic to the row type. Is it enough to just add a `case` here for histograms and set the value accordingly?

**Review by alextheimer — COMMENTED** [review]
Looks good @amolnayak311-- just a couple comments. I'll review the histogram changes once those are added, too.

File: query/src/main/scala/filodb/query/exec/SetOperatorExec.scala
Comment (alextheimer) [inline query/src/main/scala/filodb/query/exec/SetOperatorExec.scala:247]: Can you use `Double.isNan` here, instead to prevent the extra object creation?

File: query/src/main/scala/filodb/query/exec/SetOperatorExec.scala
Comment (alextheimer) [inline query/src/main/scala/filodb/query/exec/SetOperatorExec.scala:250]: Since we only set the bit when the value is not NaN, the RHS value will be returned. But Prometheus would return the NaN: http://demo.robustperception.io:9090/graph?g0.expr=vector(Nan)%20or%20sum(up)&g0.tab=1&g0.stacked=0&g0.range_input=1h

File: query/src/main/scala/filodb/query/exec/SetOperatorExec.scala
Comment (amolnayak311) [reply query/src/main/scala/filodb/query/exec/SetOperatorExec.scala:135]: Good catch but unfortunately we can't distinguish between ``vector(Nan)`` and value not being there which itself is ``NaN``

File: query/src/main/scala/filodb/query/exec/BinaryJoinExec.scala
Comment (amolnayak311) [reply query/src/main/scala/filodb/query/exec/BinaryJoinExec.scala:31]: I'm not sure I get follow the concern. Shouldn't it be adding them up?

File: query/src/main/scala/filodb/query/exec/SetOperatorExec.scala
Comment (amolnayak311) [reply query/src/main/scala/filodb/query/exec/SetOperatorExec.scala:231]: I'm not  sure any optimization is done behind the scene.

File: query/src/main/scala/filodb/query/exec/SetOperatorExec.scala
Comment (amolnayak311) [reply query/src/main/scala/filodb/query/exec/SetOperatorExec.scala:275]: Previous behavior was also incorrect. :). This behavior is consistent with AND. While I don't necessarily say this is a good thing to do. Im not sure how to deal with histograms yet and hence left a TODO

File: query/src/main/scala/filodb/query/exec/SetOperatorExec.scala
Comment (amolnayak311) [reply query/src/main/scala/filodb/query/exec/SetOperatorExec.scala:250]: Can't distinguish NaN for a missing value vs a legitimate value. I guess supporting this if required will need more changes.

File: query/src/main/scala/filodb/query/exec/SetOperatorExec.scala
Comment (alextheimer) [reply query/src/main/scala/filodb/query/exec/SetOperatorExec.scala:275]: We can leave it as long as any `or` query with a histogram operand can fail (since it wouldn't have explicitly failed before). I can't think of any example that isn't a nonsense query, though.

File: query/src/main/scala/filodb/query/exec/BinaryJoinExec.scala
Comment (alextheimer) [reply query/src/main/scala/filodb/query/exec/BinaryJoinExec.scala:31]: Huh-- I had no idea metric names were dropped during joins: http://demo.robustperception.io:9090/graph?g0.expr=node_network_up%7Bjob%3D%22node%22%7D%20%2B%20label_replace(up%7Bjob%3D%22node%22%7D%2C%22device%22%2C%22eth0%22%2C%22%22%2C%22%22)&g0.tab=1&g0.stacked=0&g0.range_input=1h

File: query/src/main/scala/filodb/query/exec/SetOperatorExec.scala
Comment (alextheimer) [reply query/src/main/scala/filodb/query/exec/SetOperatorExec.scala:231]: Super nitpick: please update this if you see this comment before the merge (I don't see this in the updates since the last review, but it's not worth calling out / blocking the merge).

**Review by alextheimer — APPROVED** [review]
LGTM as long as this is okay: https://github.com/filodb/FiloDB/pull/1669#discussion_r1351321877

**Comment by alextheimer** [comment]
If possible, please add a quick blurb for our future selves about how this same solution can (possibly) be applied to `AND` and `UNLESS`.

File: query/src/main/scala/filodb/query/exec/SetOperatorExec.scala
Comment (amolnayak311) [reply query/src/main/scala/filodb/query/exec/SetOperatorExec.scala:275]: I'll assume it is ok. We can look at any issues that may come up and fix accordingly. 
CC @vishramachandran

**Pull request merged**
